File: Implementations/Kotlin/src/main/kotlin/dev/astrolabe/protocol/inspection/RuntimeNodeDetail.kt
//
//  RuntimeNodeDetail.kt
//  astrolabe-protocol
//
//  Created by 轩辕十四 on 2026/7/20.
//

package dev.astrolabe.protocol

import kotlinx.serialization.Serializable

/** Namespaced semantic or platform attribute identifier. */
@JvmInline
@Serializable
public value class RuntimeAttributeIdentifier(
    /** Raw namespaced attribute identifier. */
    public val rawValue: String
) {
    init {
        require(RuntimeNamespacedIdentifier.isValid(rawValue)) { "Attribute identifier must be namespaced" }
    }
}

/** Namespaced category grouping related detail attributes. */
@JvmInline
@Serializable
public value class RuntimeAttributeCategory(
    /** Raw namespaced category identifier. */
    public val rawValue: String
) {
    init {
        require(RuntimeNamespacedIdentifier.isValid(rawValue)) { "Attribute category must be namespaced" }
    }
}

/** Parameters for reading one node's detail sections. */
@Serializable
public data class RuntimeNodeDetailParameters(
    /** Node whose detail sections are requested. */
    public val nodeID: RuntimeOpaqueIdentifier
) {
    public companion object {
        /** Typed request contract for the node-detail method. */
        public val contract: RuntimeMethodContract<RuntimeNodeDetailParameters> by lazy {
            RuntimeMethodContract(RuntimeMethod.nodeDetail, serializer())
        }
    }
}

/** One typed runtime attribute. */
@Serializable
public data class RuntimeAttribute(
    /** Namespaced semantic or platform attribute identifier. */
    public val identifier: RuntimeAttributeIdentifier,
    /** Typed attribute value. */
    public val value: RuntimeAttributeValue
)

/** Ordered attributes belonging to one detail category. */
@Serializable
public data class RuntimeAttributeSection(
    /** Namespaced category grouping related attributes. */
    public val category: RuntimeAttributeCategory,
    /** Attributes collected for this category. */
    public val attributes: List<RuntimeAttribute>
)

/** Successful node-detail response payload. */
@Serializable
public data class RuntimeNodeDetailPayload(
    /** Node represented by these detail sections. */
    public val nodeID: RuntimeOpaqueIdentifier,
    /** Ordered semantic and platform detail sections. */
    public val sections: List<RuntimeAttributeSection>,
    /** Optional namespaced detail metadata. */
    public val extensions: RuntimeExtensionMap? = null
) {
    public companion object {
        /** Typed success-payload contract for the node-detail method. */
        public val contract: RuntimeMethodContract<RuntimeNodeDetailPayload> by lazy {
            RuntimeMethodContract(RuntimeMethod.nodeDetail, serializer())
        }
    }
}

File: Implementations/Swift/Sources/AstrolabeProtocol/Inspection/NodeDetail/RuntimeNodeDetail.swift
//
//  RuntimeNodeDetail.swift
//  astrolabe-protocol
//
//  Created by 轩辕十四 on 2026/7/16.
//

public struct RuntimeAttributeIdentifier: Codable, Equatable, Hashable, Sendable {
    /// Namespaced attribute identifier.
    public let rawValue: String

    public init(rawValue: String) throws {
        self.rawValue = try RuntimeNamespacedIdentifier(rawValue: rawValue).rawValue
    }

    public init(from decoder: Decoder) throws {
        let container = try decoder.singleValueContainer()
        try self.init(rawValue: container.decode(String.self))
    }

    public func encode(to encoder: Encoder) throws {
        var container = encoder.singleValueContainer()
        try container.encode(rawValue)
    }
}

public struct RuntimeAttributeCategory: Codable, Equatable, Hashable, Sendable {
    /// Namespaced detail-section category.
    public let rawValue: String

    public init(rawValue: String) throws {
        self.rawValue = try RuntimeNamespacedIdentifier(rawValue: rawValue).rawValue
    }

    public init(from decoder: Decoder) throws {
        let container = try decoder.singleValueContainer()
        try self.init(rawValue: container.decode(String.self))
    }

    public func encode(to encoder: Encoder) throws {
        var container = encoder.singleValueContainer()
        try container.encode(rawValue)
    }
}

public struct RuntimeNodeDetailParameters: Codable, Equatable, Sendable, RuntimeMethodBound {
    /// Node whose detail sections are requested.
    public let nodeID: RuntimeOpaqueIdentifier

    public init(nodeID: RuntimeOpaqueIdentifier) {
        self.nodeID = nodeID
    }

    public static let runtimeMethod = RuntimeMethod.nodeDetail
}

public struct RuntimeAttribute: Codable, Equatable, Sendable {
    /// Namespaced semantic or platform attribute identifier.
    public let identifier: RuntimeAttributeIdentifier

    /// Typed attribute value.
    public let value: RuntimeAttributeValue

    public init(identifier: RuntimeAttributeIdentifier, value: RuntimeAttributeValue) {
        self.identifier = identifier
        self.value = value
    }
}

public struct RuntimeAttributeSection: Codable, Equatable, Sendable {
    /// Namespaced category grouping related attributes.
    public let category: RuntimeAttributeCategory

    /// Attributes collected for this category.
    public let attributes: [RuntimeAttribute]

    public init(category: RuntimeAttributeCategory, attributes: [RuntimeAttribute]) {
        self.category = category
        self.attributes = attributes
    }
}

public struct RuntimeNodeDetailPayload: Codable, Equatable, Sendable, RuntimeMethodBound {
    /// Node represented by these detail sections.
    public let nodeID: RuntimeOpaqueIdentifier

    /// Ordered semantic and platform detail sections.
    public let sections: [RuntimeAttributeSection]

    /// Optional namespaced detail metadata.
    public let extensions: RuntimeExtensionMap?

    public init(
        nodeID: RuntimeOpaqueIdentifier,
        sections: [RuntimeAttributeSection],
        extensions: RuntimeExtensionMap? = nil
    ) {
        self.nodeID = nodeID
        self.sections = sections
        self.extensions = extensions
    }

    public static let runtimeMethod = RuntimeMethod.nodeDetail
}
